pkg/metrics/prometheus.go

```diff
@@ -0,0 +1,152 @@
+package metrics
+
+import (
+	"github.com/grafana/grafana/pkg/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"net/http"
+	"strings"
+)
+
+const (
+	namespace = "grafana"
+	subsystem = "backend"
+)
+
+type PrometheusPublisher struct {
+	CounterCollectors map[string]*prometheus.CounterVec
+	GaugeCollectors   map[string]*prometheus.GaugeVec
+}
+
+func CreatePrometheusPublisher() (*PrometheusPublisher, error) {
+	go func() {
+		http.Handle("/metrics", prometheus.Handler())
+		err := http.ListenAndServe(":9200", nil)
```

Comment:
> Think just adding a new metrics api resource that returns data in prometheus format should be a good solution right?

That's not sufficient, Prometheus isn't designed to work with data like this. There's a difference between data being in the Prometheus format and the data being useful for Prometheus.

The Prometheus client libraries will provide good useful data for Prometheus, and allow you to also provide what other monitoring systems want. The other way doesn't generally work.

> I hate counters that never reset, terrible for anything other than rates. 

A lot of things are rates. For example in the given output, there's only 2 pieces of data being exposed which are fundamentally gauges. The other ~20 are various forms of counter.

> And rates are terrible for business/user/behavior metrics where you want accurate sums over different time periods.

If you want accuracy then you should be using transaction logs, not metrics.

Counters are more accurate than what you're currently doing, as they're much more resilient to failures. See https://www.robustperception.io/how-does-a-prometheus-counter-work/

> If prometheus can't work with count's then I guess only the timers that Grafana sends are of interest.

These have the same issues. We expect those to be sent as two counters.